corehq/apps/domain/decorators.py

```diff
@@ -58,7 +58,16 @@ def login_and_domain_required(view_func):
     @wraps(view_func)
     def _inner(req, domain, *args, **kwargs):
         user = req.user
-        domain_name, domain = load_domain(req, domain)
+        domain_identifier = domain
+        domain_name, domain = load_domain(req, domain_identifier)
+        if not domain:
+            domain = Domain.get_by_alias(domain_identifier)
+            if domain:
+                domain_name, domain = load_domain(req, domain.name)
+                return HttpResponseRedirect(
+                    req.path.replace('/' + domain_identifier + '/', '/' + domain_name + '/')
```

Comment:
You could also make sure that the string you're replacing is at the beginning, maybe with a regex replace or a `<string>.startswith(...)`